### PR TITLE
[FLINK-32610][json] JSON format supports nested type projection pushdown

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change

*This PR mainly do two aspects of work, the first is the use of Jackson streamig api parsing json, which can improve the parsing efficiency; the second is to support nested types of projection pushdown*

## Brief change log

  - *Optimize JSON schema data deserialization using jackson streaming api*
  - *JSON format supports nested type projection push down*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit tests in JsonRowDataDeSerSchemaProjectedTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by wuchong** [comment]
The build is failed.

**Comment by lsyldliu** [comment]
> * FAILURE

fixed

File: flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/canal/CanalJsonDeserializationSchema.java [hunk 1]
```diff
@@ -230,7 +230,7 @@ public void deserialize(@Nullable byte[] message, Collector<RowData> out) throws
                     return;
                 }
             }
-            final GenericRowData row = (GenericRowData) jsonDeserializer.convertToRowData(root);
+            final GenericRowData row = (GenericRowData) jsonDeserializer.deserialize(message);
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/canal/CanalJsonDeserializationSchema.java:233]: Make sense to me.

File: flink-formats/flink-json/src/test/java/org/apache/flink/formats/json/JsonParserRowDataDeSerSchemaTest.java [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.apache.flink.connector.testutils.formats.SchemaTestUtils.open;
+import static org.apache.flink.table.api.DataTypes.FIELD;
+import static org.apache.flink.table.api.DataTypes.INT;
+import static org.apache.flink.table.api.DataTypes.ROW;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Tests for {@link JsonRowDataDeserializationSchema} and {@link JsonRowDataSerializationSchema}.
```
Comment (wuchong) [inline flink-formats/flink-json/src/test/java/org/apache/flink/formats/json/JsonParserRowDataDeSerSchemaTest.java:42]: nit: `JsonRowDataDeserializationSchema` -> `JsonParserRowDataDeserializationSchema`?

File: flink-formats/flink-json/src/test/java/org/apache/flink/formats/json/JsonRowDataSerDeSchemaTest.java [hunk 3]
```diff
@@ -86,8 +89,9 @@ class JsonRowDataSerDeSchemaTest {
 
     private static final ObjectMapper OBJECT_MAPPER = JacksonMapperFactory.createObjectMapper();
 
-    @Test
-    void testSerDe() throws Exception {
+    @ParameterizedTest(name = "Is JsonParser: {0}")
```
Comment (wuchong) [inline flink-formats/flink-json/src/test/java/org/apache/flink/formats/json/JsonRowDataSerDeSchemaTest.java:92]: Why not use a class-level Parameterized test? You can take `RowLevelDeleteTest` as an example.

File: flink-formats/flink-json/src/test/java/org/apache/flink/formats/json/JsonRowDataSerDeSchemaTest.java [hunk 4]
```diff
@@ -267,8 +267,84 @@ void testSlowDeserialization() throws Exception {
         assertThat(actual).isEqualTo(expected);
     }
 
+    /**
+     * Tests parsing partial fields in actual json data, sometimes we can skip some complex field
+     * parsing, e.g., field with Row/Array type.
+     */
     @Test
-    void testSerDeMultiRows() throws Exception {
+    public void testParsePartialJson() throws Exception {
```
Comment (wuchong) [inline flink-formats/flink-json/src/test/java/org/apache/flink/formats/json/JsonRowDataSerDeSchemaTest.java:275]: Can we move this test to `JsonParserRowDataDeSerSchemaTest`?

File: flink-formats/flink-json/src/test/java/org/apache/flink/formats/json/JsonRowDataSerDeSchemaTest.java [hunk 5]
```diff
@@ -724,7 +786,7 @@ private void testParseErrors(TestSpec spec) {
                     TestSpec.json("{\"id\":\"abc\"}")
                             .rowType(ROW(FIELD("id", INT())))
                             .expectErrorMessage("Failed to deserialize JSON '{\"id\":\"abc\"}'."),
-                    TestSpec.json("{\"id\":112.013}")
+                    TestSpec.json("{\"id\":112}")
```
Comment (wuchong) [inline flink-formats/flink-json/src/test/java/org/apache/flink/formats/json/JsonRowDataSerDeSchemaTest.java:789]: Will it fail?

**Comment by lsyldliu** [comment]
@wuchong Thanks for review, I've updated.

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request merged**